File: leetcode/Python/trim-a-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def trimBST(self, root: Optional[TreeNode], low: int, high: int) -> Optional[TreeNode]:
        dq, c, p = collections.deque(), None, None
        sol = TreeNode(-1)
        sol.right = root
        dq.append([root, sol])
        while dq:
            c, p = dq.popleft()
            if c is None:
                continue
            elif c.val<low:
                if p.left == c:
                    p.left = c.right
                else:
                    p.right = c.right
                dq.append([c.right, p])
            elif c.val>high:
                if p.left == c:
                    p.left = c.left
                else:
                    p.right = c.left
                dq.append([c.left, p])
            else:
                dq.append([c.left, c])
                dq.append([c.right, c])
        return sol.right

File: leetcode/C++/trim-a-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* trimBST(TreeNode* root, int low, int high) {
        TreeNode *cur=nullptr, *p=nullptr;
        TreeNode *sol = new TreeNode(-1);
        sol->right = root;
        queue<pair<TreeNode*,TreeNode*>> q;
        q.push({root,sol});
        while(!q.empty()){
            cur=q.front().first; p=q.front().second; q.pop();
            if(cur==nullptr) continue;
            else if(cur->val<low) {
                if(p->left==cur){p->left=cur->right;}
                else {p->right=cur->right;}
                q.push({cur->right, p});
            }
            else if(cur->val>high){
                if(p->left==cur) {p->left=cur->left;}
                else {p->right=cur->left;}
                q.push({cur->left, p});
            }
            else{
                q.push({cur->left, cur});
                q.push({cur->right, cur});
            }
        }
        return sol->right;
    }
};